xblock/scorable.py

```diff
@@ -0,0 +1,143 @@
+            Return the currently saved score.  Must not calculate a new score.
+
+        ScorableXBlockMixin.set_score(self, score)
+            Set the score on the block to the specified value
+
+        ScorableXBlockMixin.calculate_score(self, score)
+            Calculate a new score based on the state of the block.  Must not
+            update any values on the block.
+    """
+
+    def rescore(self, only_if_higher):
+        """
+        Calculate a new score and save it to the block.  If only_if_higher is
+        True and the score didn't improve, keep the existing score.
+
+        Returns True if the score was changed.
+        Returns False if the score was not changed.
+        Raises a TypeError if the block cannot be scored.
+        Raises a ValueError if the user has not yet completed the problem.
+        """
+
+        _ = self.runtime.service(self, 'i18n').ugettext
+        if not self.allows_rescore():
+            self._publish_scoring_event('rescore_failure', {'failure': 'unsupported'})
+            raise TypeError(_('Problem does not support rescoring'))
+
+        original_score = self.get_score()
+        if original_score is None:
+            self._publish_scoring_event('rescore_failure', {'failure': 'unanswered'})
+            raise ValueError(_('Problem must be answered before it can be rescored.'))
+
+        try:
+            new_score = self.calculate_score()
+        except:
+            self._publish_scoring_event('rescore_failure', {'failure': 'calculation error'})
+            return False
+
+        if not only_if_higher or new_score > original_score:
+            self.set_score(new_score)
+            self._publish_scoring_event(
+                'rescore_result',
+                {
+                    'result': 'score updated',
+                    'original_score': original_score._asdict(),
+                    'new_score': new_score._asdict(),
+                }
+            )
+            return False
+        else:
+            self._publish_scoring_event(
+                'rescore_result', {
+                    'result': 'score not changed',
+                    'original_score': original_score._asdict(),
+                    'new_score': new_score._asdict(),
+                }
+            )
+            return True
+
+    def _publish_scoring_event(self, name, data):
+        """
+        Publish an event related to scoring.
+        """
+        logged = {}  # Ensure we don't mutate our arguments
+        logged.update(data)
+        logged.update(self._base_event_info())
+        self.runtime.publish(name, logged)
+
+    def _base_event_info(self):
+        """
+        Return a dictionary containing information that belongs on any logged
+        events.
+        """
+        return {
+            'usage_key': six.text_type(self.location)
+        }
+
+    def allows_rescore(self):
+        """
+        Boolean value: Can this problem be rescored.
+        """
+        return True
```

Comment:
NotImplementedError is specially used by pylint to indicate abstract methods that haven't been overridden.  It is intended to signal missing code, not runtime logic violations.  Additionally, if we try to use it for runtime checking, we'll need to silence pylint everywhere, which is less than ideal.  

We also can't just let the error bubble up, because there are reasons to disallow rescore on problem types that have those methods implemented.  The Capa XModule has the ability for the lcp to block rescoring for certain problems.  I haven't looked to deeply into why, but I assume it's used for survey problems at least.